File: ECommerceApp/Services/CartService.cs
using Microsoft.EntityFrameworkCore;
using MyECommerceApp.Models;

namespace MyECommerceApp.Services;

public class CartService
{
    private readonly ApplicationDbContext _dbContext;

    public CartService(ApplicationDbContext dbContext)
    {
        _dbContext = dbContext;
    }

    public void AddToCart(int productId, int quantity)
    {
        var item = _dbContext.CartItems.FirstOrDefault(x => x.ProductId == productId);
        if (item != null)
        {
            item.Quantity += quantity;
        }
        else
        {
            _dbContext.CartItems.Add(new CartItem { ProductId = productId, Quantity = quantity });
        }
        _dbContext.SaveChanges();
    }

    public IEnumerable<CartItem> GetCartItems()
    {
        return _dbContext.CartItems.Include(c => c.Product).ToList();
    }


    public bool RemoveFromCart(int id)
    {
        var item = _dbContext.CartItems.Find(id);
        if (item != null)
        {
            _dbContext.CartItems.Remove(item);
            _dbContext.SaveChanges();
            return true;
        }
        return false;
    }

    public bool UpdateQuantity(int id, int newQuantity)
    {
        var item = _dbContext.CartItems.Find(id);
        if (item != null)
        {
            item.Quantity = newQuantity;
            _dbContext.SaveChanges();
            return true;
        }
        return false;
    }
}

File: ECommerceApp/ClientApp/src/app/services/cart.service.ts
import {Injectable} from '@angular/core';
import {HttpClient} from '@angular/common/http';
import {Observable, throwError} from 'rxjs';
import {catchError} from 'rxjs/operators';
import {CartItem} from "../models/cart.model";
import {Product} from "../models/product.model";

@Injectable({
  providedIn: 'root'
})
export class CartService {

  constructor(private http: HttpClient) {
  }

  addToCart(product: Product): Observable<void> {
    return this.http.post<void>('https://localhost:7113/api/cart', {productId: product.id, product, quantity: 1}).pipe(
      catchError(err => {
        console.error(err);
        return throwError(() => err);
      })
    );
  }

  getCartItems(): Observable<CartItem[]> {
    return this.http.get<CartItem[]>('https://localhost:7113/api/cart').pipe(
      catchError(err => {
        console.error();
        return throwError(() => err);
      })
    );
  }

  removeFromCart(id: number): Observable<void> {
    return this.http.delete<void>(`https://localhost:7113/api/cart/${id}`).pipe(
      catchError(err => {
        console.error();
        return throwError(() => err);
      })
    );
  }

  updateQuantity(id: number, newQuantity: number): Observable<void> {
    return this.http.put<void>(`https://localhost:7113/api/cart/${id}`, {newQuantity}).pipe(
      catchError(err => {
        console.error();
        return throwError(() => err);
      })
    );
  }
}
